RFS/src/main/java/org/opensearch/migrations/bulkload/worker/DocumentsRunner.java

```diff
@@ -57,14 +85,50 @@ public CompletionStatus migrateNextShard(
                 }
             }, new IWorkCoordinator.WorkAcquisitionOutcomeVisitor<>() {
                 @Override
-                public CompletionStatus onAlreadyCompleted() throws IOException {
+                public CompletionStatus onAlreadyCompleted() {
                     return CompletionStatus.NOTHING_DONE;
                 }
 
                 @Override
                 public CompletionStatus onAcquiredWork(IWorkCoordinator.WorkItemAndDuration workItem) {
-                    doDocumentsMigration(IndexAndShardCursor.valueFromWorkItemString(workItem.getWorkItemId()), context);
-                    return CompletionStatus.WORK_COMPLETED;
+                    var docMigrationCursors = setupDocMigration(workItem.getWorkItem(), context);
+                    var latch = new CountDownLatch(1);
+                    var finishScheduler = Schedulers.newSingle( "workFinishScheduler");
+                    var disposable = docMigrationCursors
+                        .subscribeOn(finishScheduler)
+                        .doFinally(s -> finishScheduler.dispose())
+                        .takeLast(1)
+                        .subscribe(lastItem -> {},
+                            error -> log.atError()
+                                    .setCause(error)
+                                    .setMessage("Error prevented all batches from being processed")
+                                    .log(),
+                            () ->  {
+                                log.atInfo().setMessage("Reindexing completed for Index {}, Shard {}")
+                                        .addArgument(workItem.getWorkItem().getIndexName())
+                                        .addArgument(workItem.getWorkItem().getShardNumber())
+                                        .log();
+                                latch.countDown();
```

Comment:
does this get called if there's an error?  We await on the latch below - will we terminate in that case?